diff --git a/ggml/src/ggml-vulkan.cpp b/ggml/src/ggml-vulkan.cpp
index fa68360b..f9ff7b5d 100644
--- a/ggml/src/ggml-vulkan.cpp
+++ b/ggml/src/ggml-vulkan.cpp
@@ -617,9 +617,15 @@ static void ggml_vk_create_pipeline(vk_device& device, vk_pipeline& pipeline, co
         vk::PipelineCreateFlags(),
         pipeline_shader_create_info,
         pipeline->layout);
-    pipeline->pipeline = device->device.createComputePipeline(VK_NULL_HANDLE, compute_pipeline_create_info).value;
 
-    device->pipelines.push_back(pipeline);
+    try {
+        pipeline->pipeline = device->device.createComputePipeline(VK_NULL_HANDLE, compute_pipeline_create_info).value;
+        device->pipelines.push_back(pipeline);
+    } catch(vk::UnknownError const&) {
+        VK_LOG_DEBUG("Failed to create pipeline " << name);
+        ggml_vk_destroy_pipeline(device->device, pipeline);
+        pipeline.reset();
+    }
 }
 
 static void ggml_vk_destroy_pipeline(vk::Device& device, vk_pipeline& pipeline) {
